R/archive.R

```diff
@@ -196,572 +187,755 @@ next_after.Date <- function(x) x + 1L
+#'   having `version <= max_version`. The resulting object will cover a
+#'   potentially narrower `version` and `time_value` range than `x`, depending
+#'   on user-provided arguments. Otherwise, there will be one row in the output
+#'   for the `max_version` of each `time_value`. Default is `FALSE`.
+#' @importFrom data.table between key
+#' @export
+as_of.epi_archive <- function(epi_archive, max_version, min_time_value = -Inf, all_versions = FALSE) {
+  other_keys <- setdiff(
+    key(epi_archive$DT),
+    c("geo_value", "time_value", "version")
+  )
+  if (length(other_keys) == 0) other_keys <- NULL
+
+  # Check a few things on max_version
+  if (!test_set_equal(class(max_version), class(epi_archive$DT$version))) {
+    cli_abort(
+      "`max_version` must have the same classes as `epi_archive$DT$version`."
+    )
+  }
+  if (!test_set_equal(typeof(max_version), typeof(epi_archive$DT$version))) {
+    cli_abort(
+      "`max_version` must have the same types as `epi_archive$DT$version`."
+    )
+  }
+  assert_scalar(max_version, na.ok = FALSE)
+  if (max_version > epi_archive$versions_end) {
+    cli_abort("`max_version` must be at most `epi_archive$versions_end`.")
+  }
+  assert_logical(all_versions, len = 1)
+  if (!is.na(epi_archive$clobberable_versions_start) && max_version >= epi_archive$clobberable_versions_start) {
+    cli_warn(
+      'Getting data as of some recent version which could still be
+      overwritten (under routine circumstances) without assigning a new
+      version number (a.k.a. "clobbered").  Thus, the snapshot that we
+      produce here should not be expected to be reproducible later. See
+      `?epi_archive` for more info and `?epix_as_of` on how to muffle.',
+      class = "epiprocess__snapshot_as_of_clobberable_version"
+    )
+  }
+
+  # Filter by version and return
+  if (all_versions) {
+    # epi_archive is copied into result, so we can modify result directly
+    result <- epix_truncate_versions_after(epi_archive, max_version)
+    result$DT <- result$DT[time_value >= min_time_value, ]
+    return(result)
+  }
+
+  # Make sure to use data.table ways of filtering and selecting
+  as_of_epi_df <- epi_archive$DT[time_value >= min_time_value & version <= max_version, ] %>%
+    unique(
+      by = c("geo_value", "time_value", other_keys),
+      fromLast = TRUE
+    ) %>%
+    tibble::as_tibble() %>%
+    dplyr::select(-"version") %>%
+    as_epi_df(
+      geo_type = epi_archive$geo_type,
+      time_type = epi_archive$time_type,
+      as_of = max_version,
+      additional_metadata = c(epi_archive$additional_metadata,
+        other_keys = other_keys
+      )
+    )
+
+  return(as_of_epi_df)
+}
+
+
+#' @export
+fill_through_version <- function(x, ...) {
+  UseMethod("fill_through_version")
+}
+
+
+#' Fill through version
+#' @description Fill in unobserved history using requested scheme by mutating
+#'   the given object and potentially reseating its fields. See
```

Comment:
issue: notes here & probably elsewhere about mutating, mutation, and reseating need to be removed or updated.  We may still want some notes about potentially aliasing the old DT, or maybe we could just make a global warning about that in the "caution" section.

Reply:
So I've removed most of the warnings about mutation and aliasing. Two questions:
- do we actually need to warn the user about aliasing if there's no risk of mutation? seems like information we could omit and reduce mental load on the reader
- if we do decide that we want to store this somewhere, where should it go? maybe a developer doc? i added a separate note on "compactification" for archives, so we can keep the compactification arg description short and link to it from other places, and it's under a "Advanced" topic section. maybe a note on aliasing there?

Reply:
- There's a minor risk of mutation if (i) we don't encapsulate, and (ii) the user mutates.
- Advanced section sounds good.
- Eventually, post-encapsulation, and maybe if we have better interface/naming for a few things (construction from snapshots, construction from update data, extracting update data), we won't even need to mention compactification and DT and aliasing in user docs at all.  But this is like two steps down the line from this PR.

Reply:
It might also help to reduce mental load to prefix it by "For data.table users:" or something like that.